[CORE-12773] Archival: Introduce segment_collector::make_segment_upload_stream

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

Still picking apart segments under the hood, but the stream is fed by a `log_reader`.

Depends on https://github.com/redpanda-data/redpanda/pull/26676

Next: https://github.com/redpanda-data/redpanda/pull/26099

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
c/a/upload: Skip start offset adjustment for non-compacted uploads

Per 4253766df74ca8ee53e17702655976c30298ebea, this adjustment is only ever
relevant during compacted reupload, where the boundary offset of the target
range may have been compacted away.

For new segment uploads, this condition means there's a gap in the actual log.
We need to pass a valid offset range to 'async_data_uploader', so we shouldn't
perform the start_offset adjustment in this case.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
c/a/upload: Add segment_collector::make_segment_upload_stream

Still uses make_upload_candidate under the hood to collect segments but feeds
the collected offset range into segment_upload::make_segment_upload.

Returns a segment_collector_stream as before. Intended as a mostly drop-in
replacement for make_upload_candidate_stream.

Note that this is not wired all the way back to the archiver yet. Implemented
here to test the interface and keep incremental diffs relatively small and
low risk.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
c/a/archiver: Add formatters for ntp_archiver result types

Useful for debugging fixture tests

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
c/a/upload: Treat ss::semaphore_timed_out as a retryable failure

If caught during segment_upload initialization

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
c/archival: Minor clang-tidies

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
c/a/upload: Return result from align_end_offset_to_manifest

And update collector state at the call site. It's a bit easier to follow this
way, but there shouldn't be any functional difference.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d96e2877-6eba-4717-a29a-3f5d9f265201-->
<!--TEST RESULTS-->
<!--ce216086-43e2-4da5-9c1f-ae9bcf523a6c-->
<!--TEST RESULTS-->
<!--5a0ec544-c8cd-4528-8ca1-60eb5f7ec6ab-->
<!--TEST RESULTS-->
<!--af466527-1a90-449b-8b45-15ac8684b4cb-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68506">build#68506</a></summary>


| test_class                  | test_method                             | test_arguments                                                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:----------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerGroupTest           | test_group_lag_metrics                  |                                                                                                                                                                          | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68506#0197e2ab-76ec-4509-8927-888c8b0cd0db | FLAKY         | 20/21    | upstream reliability is '99.70760233918129'. current run reliability is '95.23809523809523'. drift is 4.46951 and the allowed drift is set to 50. The test should PASS |
| TieredStorageTest           | test_tiered_storage                     | {"cloud_storage_type_and_url_style": [2, "virtual_host"], "test_case": {"name": "(TS_Read == True, TS_TxRangeMaterialized == True, SpilloverManifestUploaded == True)"}} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68506#0197e2ab-ce21-4870-af8d-e97bd0505974 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RecreateTopicMetadataTest   | test_recreated_topic_metadata_are_valid | {"replication_factor": 3}                                                                                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68506#0197e2ab-ce25-451d-81f9-d81c5e7bf9d1 | FLAKY         | 19/21    | upstream reliability is '98.18913480885311'. current run reliability is '90.47619047619048'. drift is 7.71294 and the allowed drift is set to 50. The test should PASS |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test              | {"cloud_storage_type": 2}                                                                                                                                                | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68506#0197e2ab-ce25-451d-81f9-d81c5e7bf9d1 | FLAKY         | 20/21    | upstream reliability is '97.70833333333333'. current run reliability is '95.23809523809523'. drift is 2.47024 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68567">build#68567</a></summary>


| test_class                  | test_method                | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------------|:---------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RaftAvailabilityTest        | test_one_node_down         |                                                                                                                                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68567#0197e8b5-bdc2-4ddd-ac98-9c922b924ee8 | FLAKY         | 20/21    | upstream reliability is '97.58771929824562'. current run reliability is '95.23809523809523'. drift is 2.34962 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest    | test_node_operations       | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68567#0197e8b5-bdbc-45b4-b49d-1f1f1f8d6a5d | FLAKY         | 20/21    | upstream reliability is '98.72122762148338'. current run reliability is '95.23809523809523'. drift is 3.48313 and the allowed drift is set to 50. The test should PASS |
| SimpleEndToEndTest          | test_relaxed_acks          | {"write_caching": false}                                                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68567#0197e8b5-bdc2-4ddd-ac98-9c922b924ee8 | FLAKY         | 20/21    | upstream reliability is '99.25925925925925'. current run reliability is '95.23809523809523'. drift is 4.02116 and the allowed drift is set to 50. The test should PASS |
| TopicDeleteCloudStorageTest | drop_lifecycle_marker_test | {"cloud_storage_type": 2}                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68567#0197e8b3-ae83-4a3f-813d-6c06eed8bc9e | FLAKY         | 19/21    | upstream reliability is '97.47572815533981'. current run reliability is '90.47619047619048'. drift is 6.99954 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71831">build#71831</a></summary>


| test_class                 | test_method                  | test_arguments                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------|:-----------------------------|:-------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageScrubberTest   | test_scrubber                | {"cloud_storage_type": 2}                              | integration | https://buildkite.com/redpanda/redpanda/builds/71831#0199282b-9494-4967-ace6-1eeaa7ee5469 | FLAKY         | 20/21    | upstream reliability is '99.33993399339934'. current run reliability is '95.23809523809523'. drift is 4.10184 and the allowed drift is set to 50. The test should PASS |
| DatalakeClusterRestoreTest | test_slow_tiered_storage_dlq | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/71831#01992827-95e3-4f14-858b-9b633e5a0299 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71876">build#71876</a></summary>


| test_class               | test_method                           | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      |
|:-------------------------|:--------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MasterTestSuite          | test_chunk_dl_with_random_http_errors |                                                                                                                                         | unit        | https://buildkite.com/redpanda/redpanda/builds/71876#01992c86-bf91-4aea-928d-29530516b392 | FAIL          | 0/1      |                                                                                                                                                             |
| CloudTopicsTest          | test_reconciler_uploads               | {"cloud_storage_type": 2}                                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/71876#01992cb7-dcb0-4bd0-864f-6a5538063d99 | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest | test_node_operations                  | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71876#01992cb7-dcb0-41d9-be11-80f19f9f141a | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS  |
</details>
<!--/body-->

oleiman commented:
/ci-repeat 1

oleiman committed:
c/a/upload: Use log::eligible_for_compacted_reupload

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
s/test/utils: Add disk_log_builder::set_compaction_complete(index)

Helper function to ensure that a given segment looks compacted.

- mark segment as compacted
- set self_compact and clean_compact timestamps
- mark segment as finished windowed compaction

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
c/a/upload: Clean up formatters for segment_reupload types

Also adds a log line for compactedness/mode mismatch

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
c/a/tests: Add segment_collector_rpbench

Fairly basic benchmark comparing the two extant methods of producing an
upload candidate. Since the async_data_uploader-based version is neither in
use nor optimal, it's not a very interesting benchmark as written. The idea
here is to introduce a baseline as we swap out the guts of the implementation
to remove duplicated work.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented:
/ci-repeat 1

oleiman commented:
[CI Failure](https://buildkite.com/redpanda/redpanda/builds/71831#01992827-95e2-40aa-a6cf-6c5736c237c5) looks like infra

oleiman commented:
[CI Failure](https://buildkite.com/redpanda/redpanda/builds/71876#01992c86-bf91-4aea-928d-29530516b392) looks like `segment_chunk_hydration_test` timed out for some reason. Seems not related to this diff.

oleiman commented:
long live TSv2
